_posts/2021-08-16-connector-table-sql-api-part1.md

```diff
@@ -0,0 +1,234 @@
+
+1. A _factory class_ (a blueprint for creating other objects) that tells Flink with which identifier (in this case, “imap”) our connector can be addressed, which configuration options it exposes, and how the connector can be instantiated. Since Flink uses the Java Service Provider Interface (SPI) to discover factories located in different modules, you will also need to add some configuration details.
+
+2. The _table source_ object as a specific instance of the connector during the planning stage. It tells Flink some information about this instance and how it can create the connector runtime implementation. There are also more advanced features, such as [abilities](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/table/connector/source/abilities/package-summary.html), that can be implemented to improve connector performance.
+
+3. A _runtime implementation_ from the connector obtained during the planning stage. The runtime logic is implemented in Flink's core connector interfaces and does the actual work of producing rows of dynamic table data. The runtime instances are shipped to the Flink cluster. 
+
+Let us look at  this sequence (factory class → table source → runtime implementation) in reverse order.
+
+# Establish the runtime implementation of the connector
+
+You first need to have a source connector which can be used in Flink's runtime system, defining how data goes in and how it can be executed in the cluster. There are a few different interfaces available for implementing the actual source of the data and have it be discoverable in Flink.  
+
+For complex connectors, you may want to implement the [Source interface](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/api/connector/source/Source.html) which gives you a lot of control. For simpler use cases, you can use the [SourceFunction interface](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/streaming/api/functions/source/SourceFunction.html), which is the base interface for all stream data sources in Flink. There are already a few different implementations of SourceFunction interfaces for common use cases such as the [FromElementsFunction](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.html) class and the [RichSourceFunction](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/streaming/api/functions/source/RichSourceFunction.html) class. You will use the latter.  
+
+`RichSourceFunction` is a base class for implementing a parallel data source that has access to context information and some lifecycle methods. There is a `run()` method inherited from the `SourceFunction` interface that you need to implement. It is invoked once and can be used to produce the data either once for a bounded result or within a loop for an unbounded stream.
```

Comment:
I'm not sure about the description of bounded / unbounded here. Bounded really just means that the number of produced records is known to be finite, while unbounded sources can produce infinitely many records. This means bounded sources finish, while unbounded ones do not. There are more implications of this, but the point is that bounded sources can certainly run "in a loop" and for long times as well, it's just that they will eventually finish.